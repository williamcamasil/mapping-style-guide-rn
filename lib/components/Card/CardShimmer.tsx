import React, { useCallback, useMemo, useState } from 'react';
import {
  LayoutChangeEvent, StyleSheet, View,
} from 'react-native';

import { LinearGradient } from 'expo-linear-gradient';
import { createShimmerPlaceholder } from 'react-native-shimmer-placeholder';

import { useViewStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { CardSizeType, getCardSizeStyle } from './utils';

type CardShimmerPropsType = {
  size?: CardSizeType;
  theme: AppThemeType;
};

const ShimmerPlaceHolder = createShimmerPlaceholder(LinearGradient);

const styles = StyleSheet.create({
  shimmer: {
    flexDirection: 'row',
    // remove o height padr√£o da lib shimmer
    height: undefined,
  },
});

const CardShimmer: React.FC<CardShimmerPropsType> = ({
  size, theme,
}) => {
  const [shimmerWidth, setShimmerWidth] = useState(0);

  const shimeColors = useMemo(
    () => [
      theme.colors.neutralGray300,
      theme.colors.neutralGray200,
      theme.colors.neutralGray300],
    [theme],
  );

  const shimmerStyle = useViewStyles(() => [
    styles.shimmer,
    { borderRadius: theme.borders.radius.medium },
    getCardSizeStyle(theme, size),
    { padding: theme.spacings.sZero },
  ], [size, theme]);

  const onLayout = useCallback((event: LayoutChangeEvent) => {
    const { width } = event.nativeEvent.layout;
    setShimmerWidth(width);
  }, []);

  return (
    <View testID="shimmer-id" onLayout={onLayout}>
      <ShimmerPlaceHolder
        width={shimmerWidth}
        shimmerStyle={shimmerStyle}
        shimmerColors={shimeColors}
      />
    </View>
  );
};

export default withTheme(CardShimmer);

