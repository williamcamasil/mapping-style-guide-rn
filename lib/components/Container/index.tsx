import React from 'react';
import { View, ViewProps, ViewStyle } from 'react-native';

import { useViewStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';

export type ContainerPropsType = ViewProps & {
  theme: AppThemeType;
  padding?: ViewStyle['padding'];
  paddingVertical?: ViewStyle['padding'];
  paddingHorizontal?: ViewStyle['padding'];
  paddingLeft?: ViewStyle['padding'];
  paddingRight?: ViewStyle['padding'];
  paddingTop?: ViewStyle['padding'];
  paddingBottom?: ViewStyle['padding'];
  fillParent?: boolean;
  horizontal?: boolean;
  centered?: boolean;
  backgroundColor?: ViewStyle['backgroundColor'];
  blockActions?: boolean;
  justifyContent?: ViewStyle['justifyContent'];
  alignItems?: ViewStyle['alignItems'];
};

const Container: React.FC<ContainerPropsType> = ({
  children,
  style,
  theme,
  padding,
  paddingVertical,
  paddingHorizontal,
  paddingLeft,
  paddingRight,
  paddingTop,
  paddingBottom,
  fillParent,
  horizontal,
  centered,
  backgroundColor,
  blockActions,
  justifyContent,
  alignItems,
  ...others
}) => {
  const containerStyles = useViewStyles(() => [
    {
      flex: fillParent ? 1 : undefined,
      flexDirection: horizontal ? 'row' : undefined,
      justifyContent: justifyContent ?? (centered && horizontal ? 'center' : undefined),
      alignItems: alignItems ?? (centered && !horizontal ? 'center' : undefined),
      backgroundColor: backgroundColor || theme.colors.neutralWhite,
      padding,
      paddingVertical,
      paddingHorizontal,
      paddingLeft,
      paddingRight,
      paddingTop,
      paddingBottom,
    },
    style,
  ], [
    backgroundColor,
    centered,
    fillParent,
    horizontal,
    padding,
    paddingBottom,
    paddingHorizontal,
    paddingLeft,
    paddingRight,
    paddingTop,
    paddingVertical,
    style,
    theme.colors.neutralWhite,
    justifyContent,
    alignItems,
  ]);

  return (
    <View pointerEvents={blockActions ? 'none' : 'auto'} style={containerStyles} {...others}>
      {children}
    </View>
  );
};

Container.defaultProps = {
  fillParent: true,
};

export default withTheme(Container);
