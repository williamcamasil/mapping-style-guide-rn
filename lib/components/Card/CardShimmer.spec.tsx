import React from 'react';
import { StyleSheet } from 'react-native';

import { render, fireEvent, within } from '@testing-library/react-native';
import ShimmerPlaceholder from 'react-native-shimmer-placeholder';

import { DefaultTheme } from '../../theme';
import Container from '../Container';
import CardShimmer from './CardShimmer';

describe('CardShimmer', () => {

  it('Should load shimmer with active', () => {
    const result = render(<CardShimmer size="big" />);

    const shimmer = result.getByTestId('shimmer-id');
    const shimmerPlaceHolder = within(shimmer).UNSAFE_getByType(ShimmerPlaceholder);
    const shimmerStyled = StyleSheet.flatten(shimmerPlaceHolder.props.shimmerStyle);

    expect(shimmerStyled).toHaveProperty('minHeight', 128);
  });

  it('Should load shimmer with active with container and size equals to small', () => {
    const result = render(
      <Container padding={DefaultTheme.spacings.sLarge}>
        <CardShimmer size="small" />
      </Container>,
    );

    const shimmer = result.getByTestId('shimmer-id');
    const shimmerPlaceHolder = within(shimmer).UNSAFE_getByType(ShimmerPlaceholder);
    const shimmerStyled = StyleSheet.flatten(shimmerPlaceHolder.props.shimmerStyle);

    expect(shimmerStyled).toHaveProperty('width', 135);
    expect(shimmerStyled).toHaveProperty('height', 120);
  });

  it('Should load shimmer with active container and size equals to big', () => {
    const result = render(<CardShimmer size="big" />);

    const shimmer = result.getByTestId('shimmer-id');
    const shimmerPlaceHolder = within(shimmer).UNSAFE_getByType(ShimmerPlaceholder);
    const shimmerStyled = StyleSheet.flatten(shimmerPlaceHolder.props.shimmerStyle);

    expect(shimmerStyled).toHaveProperty('minHeight', 128);
  });

  it('Should loading active and verify colors to component', () => {
    const result = render(<CardShimmer size="big" />);

    const colors = [
      DefaultTheme.colors.neutralGray300,
      DefaultTheme.colors.neutralGray200,
      DefaultTheme.colors.neutralGray300,
    ];

    const shimmer = result.getByTestId('shimmer-id');
    const shimmerPlaceHolder = within(shimmer).UNSAFE_getByType(ShimmerPlaceholder);

    expect(shimmerPlaceHolder.props.shimmerColors).toEqual(colors);
  });

  it('Should load shimmer and call onLayout event to load width', () => {
    const result = render(<CardShimmer size="big" />);

    const shimmer = result.getByTestId('shimmer-id');

    fireEvent(shimmer, 'layout', {
      nativeEvent: { layout: { width: 300 } },
    });

    const shimmerPlaceHolder = within(shimmer).UNSAFE_getByType(ShimmerPlaceholder);

    expect(shimmerPlaceHolder.props).toHaveProperty('width', 300);
  });
});

describe('snapshot', () => {
  it('Should loading active with container and size equals to small', () => {
    const result = render(
      <Container padding={DefaultTheme.spacings.sLarge}>
        <CardShimmer size="small" />
      </Container>,
    );

    expect(result.toJSON()).toMatchSnapshot();
  });

  it('Should loading active with container and size equals to big', () => {
    const result = render(<CardShimmer size="big" />);

    expect(result.toJSON()).toMatchSnapshot();
  });
});
