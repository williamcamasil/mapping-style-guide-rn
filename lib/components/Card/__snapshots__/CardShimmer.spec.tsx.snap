// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Should loading active with container and size equals to big 1`] = `
<View
  onLayout={[Function]}
  testID="shimmer-id"
>
  <View
    style={
      [
        {
          "height": 15,
          "width": 0,
        },
        {
          "overflow": "hidden",
        },
        {
          "borderRadius": 12,
          "flexDirection": "row",
          "height": undefined,
          "minHeight": 128,
          "padding": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "height": 0,
            "opacity": 0,
            "width": 0,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        {
          "backgroundColor": "#E1E2E4",
          "flex": 1,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "transform": [
              {
                "translateX": -0,
              },
            ],
          }
        }
      >
        <View
          colors={
            [
              "#E1E2E4",
              "#F0F0F2",
              "#E1E2E4",
            ]
          }
          end={
            {
              "x": 2,
              "y": 0.5,
            }
          }
          locations={
            [
              0.3,
              0.5,
              0.7,
            ]
          }
          start={
            {
              "x": -1,
              "y": 0.5,
            }
          }
          style={
            {
              "flex": 1,
              "width": 0,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`snapshot Should loading active with container and size equals to small 1`] = `
<View
  pointerEvents="auto"
  style={
    {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": undefined,
      "justifyContent": undefined,
      "padding": 28,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
    }
  }
>
  <View
    onLayout={[Function]}
    testID="shimmer-id"
  >
    <View
      style={
        [
          {
            "height": 15,
            "width": 0,
          },
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 12,
            "flexDirection": "row",
            "height": 120,
            "padding": 0,
            "width": 135,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "height": 0,
              "opacity": 0,
              "width": 0,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          {
            "backgroundColor": "#E1E2E4",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "flex": 1,
              "transform": [
                {
                  "translateX": -0,
                },
              ],
            }
          }
        >
          <View
            colors={
              [
                "#E1E2E4",
                "#F0F0F2",
                "#E1E2E4",
              ]
            }
            end={
              {
                "x": 2,
                "y": 0.5,
              }
            }
            locations={
              [
                0.3,
                0.5,
                0.7,
              ]
            }
            start={
              {
                "x": -1,
                "y": 0.5,
              }
            }
            style={
              {
                "flex": 1,
                "width": 0,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
